Fixed #32902 -- Fixed CsrfViewMiddleware.process_response()'s cookie reset logic.

cjerdonek opened the pull request:
https://code.djangoproject.com/ticket/32902

shaib reviewed (commented):
At a glimpse, the changes look OK, but I couldn't give this proper attention at the moment. I intend to return to this over the weekend.

cjerdonek commented:
Thanks a lot, @shaib. Let me know if you have any questions.

shaib commented on django/middleware/csrf.py:
```diff
@@ -437,15 +434,15 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
-                return response
-
-        if not request.META.get("CSRF_COOKIE_USED", False):
+        # Prevent _set_token() from being called twice.
```
`_set_token()` is only called here (below). This is within `process_response()`. The only reasonable case where it would be called twice is when both middleware and decorator are used (the less reasonable case is when the decorator is somehow applied more than once to the same view). I'd add to the comment something like 
```suggestion
        # Prevent _set_token() from being called twice (e.g. when both decorator and middleware used)
```

cjerdonek replied:
Thanks, @shaib. I updated the comment with your suggestion.

shaib reviewed (approved):
This simplifies a little and, AFAICT, keeps functionality intact. One small comment about a comment, not critical; I think this is good to go.

felixxm commented on django/middleware/csrf.py:
```diff
@@ -106,10 +106,7 @@ def rotate_token(request):
-        "CSRF_COOKIE_USED": True,
-        "CSRF_COOKIE": _get_new_csrf_token(),
-    })
+    request.META["CSRF_COOKIE"] = _get_new_csrf_token()
```
I would remove the last commit, this is not necessary for this ticket. Moreover, as far as I'm aware ticket-32916 suggests to do the opposite and remove rather `csrf_cookie_needs_reset`.

cjerdonek replied:
Okay, I'll remove it. I added this commit because it's related to the PR: `rotate_token()` doesn't imply `"CSRF_COOKIE_USED"` but does imply needing a reset. (The ticket suggests using a `"CSRF_COOKIE_NEEDS_RESET"` key rather than `"CSRF_COOKIE_USED"` because it's the former that is named appropriately for all cases.)

felixxm replied:
Right, I missed that the ticket suggest a different name. I would still leave it for ticket-32916.

cjerdonek replied:
Okay, I removed the commit. (If [ticket-32916](https://code.djangoproject.com/ticket/32916) isn't accepted, I still think the commit should be considered separately for the reason I stated above -- that only cookie needs reset is true in this case.)

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -437,15 +437,16 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
-        if not request.META.get("CSRF_COOKIE_USED", False):
+        # Prevent _set_token() from being called twice (e.g. when both a
+        # decorator and middleware are used).
+        if getattr(response, 'csrf_cookie_set', False):
```
I cannot think of any legitimate reason on how you could trigger this sensibly to go wrong, but do imagine the following (order):

 * csrf middleware
 * custom middleware
 * decorator 
 
 Assuming "custom middleware" calls `rotate_token` (in `process_response` or so) which sets `csrf_cookie_needs_reset` to True but the decorator already did set `csrf_cookie_set` to True as well, then the token wouldn't get updated.
 
 The code (not nice, but I couldn't come up with anything better to express what I ment) would look like this:
 ```py
     def process_response(self, request, response):
        # The cookie needs to be reset, reset it but also reset the flag so we do it not more often than neccessary (can still be more than once per request)
        if getattr(request, 'csrf_cookie_needs_reset', False):
            self.csrf_cookie_needs_reset = False
        elif getattr(response, 'csrf_cookie_set', False):
            return response
        elif not request.META.get("CSRF_COOKIE_USED", False):
            return response

        # Set the CSRF cookie even if it's already set, so we renew
        # the expiry timer.
        self._set_token(request, response)
        response.csrf_cookie_set = True
        return response
```

Not trying to bike-shed here, I honestly do not know how `process_response` is supposed to behave in such edge-cases (or whether they even make sense at all)

cjerdonek replied:
That's a great thought, @apollo13. Thanks!

I will note that prior to PR #5605 (so prior to Django 1.10), I don't think the scenario you describe would have been supported. See below for what `process_response()` looked like at that time (where `csrf_processing_done` is playing the role of `csrf_cookie_set`). The logic back then is basically what I'm proposing now:
https://github.com/django/django/blob/6d9c5d46e644a8ef93b0227fc710e09394a03992/django/middleware/csrf.py#L221-L241

However, I don't see any reason why we can't support that new scenario. Let me add a test case for that situation, and we can see how it looks.

apollo13 commented:
+1 for the test case. Just to clarify: I (and probably others as well) have no idea on what is supposed to be supported. That thing organically grew to the point where those questions do arise. So my main questions are probably: does it make sense to support it from an end user perspective. And if yes is the added (?) complexity worth to maintain it?

On Wed, Jul 14, 2021, at 08:54, Chris Jerdonek wrote:
> 
> 
> ***@***.**** commented on this pull request.
> 
> In django/middleware/csrf.py 
> <https://github.com/django/django/pull/14599#discussion_r669340117>:
> 
> > @@ -437,15 +437,16 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
>          return self._accept(request)
>  
>      def process_response(self, request, response):
> -        if not getattr(request, 'csrf_cookie_needs_reset', False):
> -            if getattr(response, 'csrf_cookie_set', False):
> -                return response
> -
> -        if not request.META.get("CSRF_COOKIE_USED", False):
> +        # Prevent _set_token() from being called twice (e.g. when both a
> +        # decorator and middleware are used).
> +        if getattr(response, 'csrf_cookie_set', False):
> That's a great thought, @apollo13 <https://github.com/apollo13>. Thanks!
> 
> I will note that prior to PR #5605 
> <https://github.com/django/django/pull/5605> (so prior to Django 1.10), 
> I don't think the scenario you describe would have been supported. See 
> below for what `process_response()` looked like at that time (where 
> `csrf_processing_done` is playing the role of `csrf_cookie_set`). The 
> logic back then is basically what I'm proposing now:
> https://github.com/django/django/blob/6d9c5d46e644a8ef93b0227fc710e09394a03992/django/middleware/csrf.py#L221-L241
> 
> However, I don't see any reason why we can't support that new scenario. 
> Let me add a test case for that situation, and we can see how it looks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/django/django/pull/14599#discussion_r669340117>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAT5C4XGWWGCTRB2Q64JHLTXUYBFANCNFSM473S25RQ>.
>

cjerdonek commented:
Okay, I updated the PR. I added a test to cover the scenario that @apollo13 described, and I wrote the test so it could be included in the mixin test class so that it is tested under both `CSRF_USE_SESSIONS` true and false. The test was added in the first commit in the sequence, and it passes both before and after the code change.

I also made the changes to `process_response()` along the lines @apollo13 suggested, with comments adding additional explanation. With the changes, the two failing tests pass, and the new passing test continues to pass.

cjerdonek commented:
> So my main questions are probably: does it make sense to support it from an end user perspective. And if yes is the added (?) complexity worth to maintain it?

To respond to @apollo13's question, I do think it's worth adding explicit support for and testing this scenario (calling `rotate_token()` inside `process_response()`). Some reasons are: (1) the scenario is already working today (it's just not tested). (2) If it stops working, it will have a bad security effect, namely people will think tokens are getting rotated but they won't be, and this won't necessarily be easy to notice. (3) It's not _that_ much complexity (just an extra conditional or so in `process_response()`). (4) This extra complexity (and more) should all go away once [ticket #32916](https://code.djangoproject.com/ticket/32916) is addressed anyways, leaving just the code that tests the scenario. (I'm waiting for this PR to be resolved before submitting the PR for that ticket.) (5) In the alternative of not supporting this scenario, we could address (2) by raising an exception or warning in that case, but that just adds a different type of extra complexity (and more than what would be left after (4)).

felixxm commented:
@cjerdonek Thanks for updates :+1:
